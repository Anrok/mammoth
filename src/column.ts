import { InternalExpression, makeExpression } from './expression';
import { StringToken, Token } from './tokens';

import { Table } from './table';
import { toSnakeCase } from './naming/snake-case';

export interface ColumnDefinition<DataType, IsNotNull extends boolean, HasDefault extends boolean> {
  notNull(): ColumnDefinition<DataType, true, HasDefault>;
  primaryKey(): ColumnDefinition<DataType, true, HasDefault>;
  // In most cases a default clause means you do not need to provide any value during insert. In
  // theory however it's possible the default expression doesn't set a value in some case. In the
  // case of a NOT NULL constraint this would mean you'd have to set a value when inserting. Because
  // this is not neccesary in most of the cases we just assume a default expression will always set
  // a value. You can opt out of this by setting `IsAlwaysSettingAValue` to false.
  default<IsAlwaysSettingAValue extends boolean = true>(
    defaultExpression: string
  ): ColumnDefinition<DataType, IsNotNull, IsAlwaysSettingAValue>;
  check(): ColumnDefinition<DataType, IsNotNull, HasDefault>;
  unique(): ColumnDefinition<DataType, IsNotNull, HasDefault>;
  references<
    T extends Table<any, any>,
    ColumnName extends keyof (T extends Table<any, infer Columns> ? Columns : never)
  >(
    table: T,
    columnName: ColumnName
  ): ColumnDefinition<DataType, IsNotNull, HasDefault>;
}

export const makeColumnDefinition = <
  DataType,
  IsNotNull extends boolean,
  HasDefault extends boolean
>(
  dataType: string
): ColumnDefinition<DataType, IsNotNull, HasDefault> => {
  return {
    notNull() {
      return this as any;
    },

    primaryKey() {
      return this as any;
    },

    default(defaultExpression) {
      return this as any;
    },

    check() {
      return this as any;
    },

    unique() {
      return this as any;
    },

    references(table, columnName) {
      return this as any;
    },
  };
};

export interface Column<
  Name,
  TableName,
  DataType,
  IsNotNull extends boolean,
  HasDefault extends boolean,
  JoinType
> extends InternalExpression<Name, DataType, IsNotNull> {
  _columnBrand: any;

  as<AliasName extends string>(
    alias: AliasName
  ): Column<AliasName, TableName, DataType, IsNotNull, HasDefault, JoinType>;

  /** @internal */
  toTokens(): Token[];
}

export const makeColumn = <
  ColumnName,
  TableName,
  DataType,
  IsNotNull extends boolean,
  HasDefault extends boolean
>(
  columnName: ColumnName,
  tableName: TableName,
  originalColumnName: string | undefined
): Column<ColumnName, TableName, DataType, IsNotNull, HasDefault, undefined> => {
  // TODO: either use some sort of AliasToken which we can later strip. Or pass some sort of option
  // to the token generation so we can choose if we want named or unnamed.

  return {
    _columnBrand: undefined,

    ...makeExpression((aliasBehaviour) => {
      const snakeCaseColumnName = toSnakeCase((columnName as unknown) as string);

      if (aliasBehaviour === `INCLUDE`) {
        return originalColumnName
          ? [new StringToken(`${tableName}.${toSnakeCase(originalColumnName)} "${columnName}"`)]
          : [
              snakeCaseColumnName === (columnName as unknown)
                ? new StringToken(`${tableName}.${snakeCaseColumnName}`)
                : new StringToken(`${tableName}.${snakeCaseColumnName} "${columnName}"`),
            ];
      }

      return originalColumnName
        ? [new StringToken(`${tableName}.${toSnakeCase(originalColumnName)}`)]
        : [new StringToken(`${tableName}.${snakeCaseColumnName}`)];
    }),

    as(alias) {
      return makeColumn(alias, tableName, (columnName as unknown) as string);
    },

    toTokens() {
      const snakeCaseColumnName = toSnakeCase((columnName as unknown) as string);

      return originalColumnName
        ? [new StringToken(`${tableName}.${toSnakeCase(originalColumnName)} "${columnName}"`)]
        : [
            snakeCaseColumnName === (columnName as unknown)
              ? new StringToken(`${tableName}.${snakeCaseColumnName}`)
              : new StringToken(`${tableName}.${snakeCaseColumnName} "${columnName}"`),
          ];
    },
  };
};
