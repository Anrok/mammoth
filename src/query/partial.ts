import { GroupToken, ParameterToken, StringToken, Token, NoSpaceToken } from '../tokens';
import { Tokenable, Query } from './base';

// TODO: we can probably change this to be an expression.
export class PartialQuery implements Tokenable {
  tokens: Token[] = [];

  constructor(...tokens: Token[]) {
    this.tokens = tokens || [];
  }

  toTokens() {
    return this.tokens;
  }

  private add(partialQuery: PartialQuery | Query<any, any, any>, separator: string) {
    this.tokens.push(new StringToken(separator), ...partialQuery.toTokens());
    return this;
  }

  or(partialQuery: PartialQuery) {
    return this.add(partialQuery, `OR`);
  }

  and(partialQuery?: PartialQuery) {
    this.tokens.push(new StringToken(`AND`));

    if (partialQuery) {
      this.tokens.push(...partialQuery.toTokens());
    }
    return this;
  }

  andNotExists(partialQuery: PartialQuery | Query<any, any, any>) {
    this.tokens.push(new StringToken(`AND NOT EXISTS`), new GroupToken(partialQuery.toTokens()));
    return this;
  }

  andExists(partialQuery: PartialQuery | Query<any, any, any>) {
    this.tokens.push(new StringToken(`AND EXISTS`), new GroupToken(partialQuery.toTokens()));
    return this;
  }

  interval() {
    return this.castTo(`interval`);
  }

  castTo(type: string) {
    // TODO: should we enclose the already existing tokens in a group token so we surround it with
    // ( and ).
    this.tokens.push(new NoSpaceToken(`::${type}`));
    return this;
  }

  minus(string: string | PartialQuery) {
    if (string instanceof PartialQuery) {
      this.tokens.push(new StringToken(`-`), ...string.toTokens());
    } else {
      this.tokens.push(new StringToken(`-`), new ParameterToken(string));
    }

    return this;
  }

  plus(string: string) {
    this.tokens.push(new StringToken(`+`), new ParameterToken(string));
    return this;
  }

  gt(partialQuery: PartialQuery) {
    this.tokens.push(new StringToken(`>`), ...partialQuery.toTokens());
    return this;
  }

  lt(partialQuery: PartialQuery) {
    this.tokens.push(new StringToken(`<`), ...partialQuery.toTokens());
    return this;
  }

  gte(partialQuery: PartialQuery) {
    this.tokens.push(new StringToken(`>=`), ...partialQuery.toTokens());
    return this;
  }

  lte(partialQuery: PartialQuery) {
    this.tokens.push(new StringToken(`<=`), ...partialQuery.toTokens());
    return this;
  }

  not(partialQuery?: PartialQuery) {
    this.tokens.push(new StringToken(`NOT`));
    if (partialQuery) {
      this.tokens.push(...partialQuery.toTokens());
    }
    return this;
  }

  in(object: any[] | PartialQuery | Query<any, any, any>) {
    if (object instanceof PartialQuery) {
      this.tokens.push(new StringToken(`IN`), new GroupToken(object.tokens));
    } else if (object instanceof Query) {
      this.tokens.push(new StringToken(`IN`), new GroupToken(object.toTokens()));
    } else {
      this.tokens.push(new StringToken(`IN`), new GroupToken([new ParameterToken(object)]));
    }
    return this;
  }

  // TODO: Move this to an order by query.
  nullsFirst() {
    this.tokens.push(new StringToken(`NULLS FIRST`));
    return this;
  }

  nullsLast() {
    this.tokens.push(new StringToken(`NULLS LAST`));
    return this;
  }
}
