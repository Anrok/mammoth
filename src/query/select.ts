import { ColumnWrapper } from '../columns';
import { Database } from '../database';
import { Table } from '../table';
import { CollectionToken, SeparatorToken, StringToken } from '../tokens';
import { Tokenable, Query, QueryType } from './base';

export class SelectQuery<
  Db extends Database<any>,
  Row,
  InsertRow,
  UpdateRow,
  Ret,
  SingleRet,
  Tables = undefined
> extends Query<Db, Ret, SingleRet, Tables> {
  protected type: QueryType = 'ROWS';

  /** @internal */
  from<T extends Table<any, any, any>>(
    table: T,
  ): SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables & T['$row']> {
    // TODO: move this to the constructor so we do not have to expose SelectQuery#from().

    return this.internalFrom(table);
  }

  join<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'JOIN',
      table,
    );
  }

  crossJoin<T extends Table<any>>(table: T) {
    this.tokens.push(new StringToken(`CROSS JOIN`), new StringToken(table.getName()));
    return this;
  }

  innerJoin<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'INNER JOIN',
      table,
    );
  }

  leftJoin<T extends Table<any>>(table: T) {
    // TODO: this isn't rock solid yet. It's changing the return type of the columns selected from the
    // left joined table to be undefined (as that's how it supposed to be). But, we're targeting the
    // columns based on the name, so a collision is likely and column#as() is not supported as it doesn't
    // change the return type.
    return this.internalJoin<
      T,
      SelectQuery<
        Db,
        Row,
        InsertRow,
        UpdateRow,
        // @ts-ignore
        (Pick<Ret[number], Exclude<keyof Ret[number], keyof T['$row']>> &
          Pick<
            { [P in keyof T['$row']]?: T['$row'][P] },
            // @ts-ignore
            Extract<keyof Ret[number], keyof T['$row']>
          >)[],
        Pick<SingleRet, Exclude<keyof SingleRet, keyof T['$row']>> &
          Pick<
            { [P in keyof T['$row']]?: T['$row'][P] },
            Extract<keyof SingleRet, keyof T['$row']>
          >,
        Tables
      >
    >('LEFT JOIN', table);
  }

  rightJoin<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'RIGHT JOIN',
      table,
    );
  }

  leftOuterJoin<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'LEFT OUTER JOIN',
      table,
    );
  }

  rightOuterJoin<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'RIGHT OUTER JOIN',
      table,
    );
  }

  fullOuterJoin<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'FULL OUTER JOIN',
      table,
    );
  }

  fullJoin<T extends Table<any>>(table: T) {
    return this.internalJoin<T, SelectQuery<Db, Row, InsertRow, UpdateRow, Ret, SingleRet, Tables>>(
      'FULL JOIN',
      table,
    );
  }

  where(tokenable: Tokenable) {
    return this.internalWhere(tokenable);
  }

  limit(limit: number | 'ALL') {
    this.append`LIMIT`;

    if (typeof limit === 'number') {
      this.tokens.push(new StringToken(String(limit)));
    } else {
      this.tokens.push(new StringToken(`ALL`));
    }

    return this;
  }

  offset(offset: number) {
    this.tokens.push(new StringToken(`OFFSET ${offset}`));
    return this;
  }

  having(tokenable: Tokenable) {
    // TODO: should tokenable be a list like in orderBy?

    this.tokens.push(new StringToken(`HAVING`), ...tokenable.toTokens());
    return this;
  }

  orderBy(...tokenables: Tokenable[]) {
    this.tokens.push(
      new StringToken(`ORDER BY`),
      new SeparatorToken(
        `,`,
        tokenables.map(tokenable => new CollectionToken(tokenable.toTokens())),
      ),
    );
    return this;
  }

  forUpdate() {
    this.tokens.push(new StringToken(`FOR UPDATE`));
    return this;
  }

  forNoKeyUpdate() {
    this.tokens.push(new StringToken(`FOR UPDATE`));
    return this;
  }

  forShare() {
    this.tokens.push(new StringToken(`FOR UPDATE`));
    return this;
  }

  forNoKeyShare() {
    this.tokens.push(new StringToken(`FOR UPDATE`));
    return this;
  }

  of(...tables: Table<any>[]) {
    this.tokens.push(
      new StringToken(`OF`),
      new SeparatorToken(
        `,`,
        tables.map(table => new StringToken(table.getName())),
      ),
    );
    return this;
  }

  skipLocked() {
    this.tokens.push(new StringToken(`SKIP LOCKED`));
    return this;
  }

  nowait() {
    this.tokens.push(new StringToken(`NOWAIT`));
    return this;
  }

  groupBy(...columns: ColumnWrapper<any, any, any, any, any>[]) {
    this.tokens.push(
      new StringToken(`GROUP BY`),
      new SeparatorToken(
        `,`,
        columns.map(column => new CollectionToken(column!.toTokens())),
      ),
    );
    return this;
  }
}
