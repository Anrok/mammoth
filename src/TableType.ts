import { Column } from './column';
import { DbConfig } from './config';

export type AnyTable = Table<any, any, any>;

export type Table<Config extends DbConfig, TableName, Columns> = Columns &
  InternalTable<Config, TableName, Columns>;

export interface InternalTable<Config extends DbConfig, TableName, Columns> {
  /** @internal */
  _tableBrand: any;

  /** @internal */
  getName(): string;

  /** @internal */
  getOriginalName(): string;

  // Because we use the column's table name to determine whether the data type should be nullable
  // when joining, we change the column's table name to the alias.
  as<T>(alias: T): Table<
    Config,
    T,
    {
      [K in keyof Columns]: Columns[K] extends Column<
        Config,
        infer Name,
        string,
        infer DataType,
        infer IsNotNull,
        infer HasDefault,
        infer JoinType
      >
        ? Column<Config, Name, T, DataType, IsNotNull, HasDefault, JoinType>
        : never;
    }
  >;
}
