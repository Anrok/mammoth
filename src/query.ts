// This is used so we can capture the Returning of every different query type without having to
// specify every query seperately. The private property is used to simulate a nominal type so only
import { DbConfig, GetResultType } from './config';

import { ResultSet } from './result-set';
import { Token } from './tokens';

export type SpecificQuery<
  Config extends DbConfig,
  DataType,
  Q extends Query<any>,
  Result = ResultSet<Config, Q, false>,
> = Result extends {
  [K in keyof Result]: GetResultType<Config, DataType> | GetResultType<Config, 'Null'>;
}
  ? Q
  : never;

export type BooleanQuery<Config extends DbConfig, Q extends Query<any>> = SpecificQuery<
  Config,
  boolean,
  Q
>;

export type AnyQuery = Query<any>;

// this class is captured when doing a conditional type check (through T extends Query<infer Returning>).
export abstract class Query<Returning> {
  private _queryBrand!: ['query', Returning];
  /** @internal */
  abstract toTokens(includeAlias?: boolean): Token[];
  /** @internal */
  abstract getReturningKeys(): string[];
}
