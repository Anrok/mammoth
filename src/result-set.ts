import type { Column } from './column';
import { DeleteQuery } from './delete';
import type { Expression } from './expression';
import { InsertQuery } from './insert';
import { Query } from './query';
import { SelectQuery } from './select';
import { UpdateQuery } from './update';

export type ResultSetDataType<Type, IsNotNull extends boolean> = IsNotNull extends true ? Type : Type | null;

// This is not ideal, but, using dts-jest and it's snapshotting it's not capable to snapshot an e.g.
// optional number to `number | undefined`. Instead, it will snapshot to `number`. Because it's
// important to get the optional behaviour under test as well (it's so easy to create a regression)
// this flag is introduced to return a nominal class which gets snapshotted with the correct info.
export type ResultSet<T extends Query<any>> = T extends SelectQuery<
  infer Returning
>
  ? {
      [K in keyof Returning]: Returning[K] extends Column<
        any,
        any,
        infer D,
        infer N,
        any,
        infer JoinType
      >
        ? Extract<JoinType, 'left-join'> extends never
          ? Extract<JoinType, 'left-side-of-right-join'> extends never
            ? Extract<JoinType, 'full-join'> extends never
              ? N extends true
                ? ResultSetDataType<D, true>
                : ResultSetDataType<D, false>
              : ResultSetDataType<D, false>
            : ResultSetDataType<D, false>
          : ResultSetDataType<D, false>
        : Returning[K] extends Expression<infer D, infer IsNotNull, any>
        ? ResultSetDataType<D, IsNotNull>
        : Returning[K] extends Query<{}>
        ? ResultSet<Returning[K]>[keyof ResultSet<Returning[K]>]
        : never;
    }
  : T extends DeleteQuery<any, infer Returning>
  ? {
      [K in keyof Returning]: Returning[K] extends Column<
        any,
        any,
        infer D,
        infer N,
        any,
        infer JoinType
      >
        ? Extract<JoinType, 'left-join'> extends never
          ? Extract<JoinType, 'left-side-of-right-join'> extends never
            ? Extract<JoinType, 'full-join'> extends never
              ? N extends true
                ? ResultSetDataType<D, true>
                : ResultSetDataType<D, false>
              : ResultSetDataType<D, false>
            : ResultSetDataType<D, false>
          : ResultSetDataType<D, false>
        : Returning[K] extends Expression<infer D, infer IsNotNull, any>
        ? ResultSetDataType<D, IsNotNull>
        : Returning[K] extends Query<{}>
        ? ResultSet<Returning[K]>[keyof ResultSet<Returning[K]>]
        : never;
    }
  : T extends UpdateQuery<any, infer Returning>
  ? {
      [K in keyof Returning]: Returning[K] extends Column<
        any,
        any,
        infer D,
        infer N,
        any,
        infer JoinType
      >
        ? Extract<JoinType, 'left-join'> extends never
          ? Extract<JoinType, 'left-side-of-right-join'> extends never
            ? Extract<JoinType, 'full-join'> extends never
              ? N extends true
                ? ResultSetDataType<D, true>
                : ResultSetDataType<D, false>
              : ResultSetDataType<D, false>
            : ResultSetDataType<D, false>
          : ResultSetDataType<D, false>
        : Returning[K] extends Expression<infer D, infer IsNotNull, any>
        ? ResultSetDataType<D, IsNotNull>
        : Returning[K] extends Query<{}>
        ? ResultSet<Returning[K]>[keyof ResultSet<Returning[K]>]
        : never;
    }
  : T extends InsertQuery<any, infer Returning>
  ? {
      [K in keyof Returning]: Returning[K] extends Column<
        any,
        any,
        infer D,
        infer N,
        any,
        infer JoinType
      >
        ? Extract<JoinType, 'left-join'> extends never
          ? Extract<JoinType, 'left-side-of-right-join'> extends never
            ? Extract<JoinType, 'full-join'> extends never
              ? N extends true
                ? ResultSetDataType<D, true>
                : ResultSetDataType<D, false>
              : ResultSetDataType<D, false>
            : ResultSetDataType<D, false>
          : ResultSetDataType<D, false>
        : Returning[K] extends Expression<infer D, infer IsNotNull, any>
        ? ResultSetDataType<D, IsNotNull>
        : Returning[K] extends Query<{}>
        ? ResultSet<Returning[K]>[keyof ResultSet<Returning[K]>]
        : never;
    }
  : never;
