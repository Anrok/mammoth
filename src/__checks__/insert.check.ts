import {
  defineDb,
  defineTable,
  integer,
  raw,
  serial,
  text,
  timestampWithTimeZone,
  uuid,
} from '../';
import { Table } from '../TableType';
import { Column } from '../column';
import { InsertQuery } from '../insert';
import { Query } from '../query';
import { ResultSet } from '../result-set';
import { expect, describe, test } from 'tstyche';

const toSnap = <T extends Query<any>>(query: T): ResultSet<T> => {
  return undefined as any;
};

const foo = defineTable({
  id: uuid().primaryKey().default(`gen_random_uuid()`),
  createDate: timestampWithTimeZone().notNull().default(`now()`),
  name: text().notNull(),
  value: integer(),
});

const serialTest = defineTable({
  id: serial().primaryKey(),
  value: integer(),
});

const db = defineDb({ foo, serialTest }, () => Promise.resolve({ rows: [], affectedCount: 0 }));

describe('insert', () => {
  test('should insert and returning count', () => {
    expect(toSnap(db.insertInto(db.foo).values({ name: `Test` }))).type.toBeNumber();
  });

  test('should insert multiple rows and returning count', () => {
    expect(
      toSnap(db.insertInto(db.foo).values([{ name: `Test` }, { name: `Test 2` }])),
    ).type.toBeNumber();
  });

  test('should insert default column', () => {
    expect(
      toSnap(db.insertInto(db.foo).values({ name: `Test`, createDate: new Date() })),
    ).type.toBeNumber();
  });

  test('should not insert unknown column', () => {
    expect(toSnap(db.insertInto(db.foo).values({ name: `Test`, asd: `Test` }))).type.toRaiseError();
  });

  test('should not insert invalid type in known column', () => {
    expect(toSnap(db.insertInto(db.foo).values({ name: 123 }))).type.toRaiseError();
  });

  test('should not insert multiple rows with invalid colums', () => {
    expect(
      toSnap(db.insertInto(db.foo).values([{ name: `Test` }, { name: `Test 2`, asd: 123 }])),
    ).type.toRaiseError();
  });

  test('should insert and await affect count', async () => {
    expect(await db.insertInto(db.foo).values({ name: `Test` })).type.toBeNumber();
  });

  test('should insert-returning and await rows', async () => {
    expect(await db.insertInto(db.foo).values({ name: `Test` }).returning(`name`)).type.toBe<
      { name: string }[]
    >();
  });

  test('should insert without explicit value for column serial', () => {
    expect(db.insertInto(db.serialTest).values({ value: 123 })).type.toBe<
      InsertQuery<
        Table<
          'serialTest',
          {
            id: Column<'id', 'serialTest', number, true, true, undefined>;
            value: Column<'value', 'serialTest', number, false, false, undefined>;
          }
        >,
        number,
        {
          id: Column<'id', 'serialTest', number, true, true, undefined>;
          value: Column<'value', 'serialTest', number, false, false, undefined>;
        }
      >
    >();
  });

  test('should insert with expression of the not-null correct type', () => {
    expect(
      db.insertInto(db.serialTest).values({ value: raw<number, true>`get_value()` }),
    ).type.toBe<
      InsertQuery<
        Table<
          'serialTest',
          {
            id: Column<'id', 'serialTest', number, true, true, undefined>;
            value: Column<'value', 'serialTest', number, false, false, undefined>;
          }
        >,
        number,
        {
          id: Column<'id', 'serialTest', number, true, true, undefined>;
          value: Column<'value', 'serialTest', number, false, false, undefined>;
        }
      >
    >();
  });

  test('should insert with expression of the nullable correct type', () => {
    expect(
      db.insertInto(db.serialTest).values({ value: raw<number, false>`get_value()` }),
    ).type.toBe<
      InsertQuery<
        Table<
          'serialTest',
          {
            id: Column<'id', 'serialTest', number, true, true, undefined>;
            value: Column<'value', 'serialTest', number, false, false, undefined>;
          }
        >,
        number,
        {
          id: Column<'id', 'serialTest', number, true, true, undefined>;
          value: Column<'value', 'serialTest', number, false, false, undefined>;
        }
      >
    >();
  });

  test('should not insert with wrong type of expression', () => {
    expect(
      db.insertInto(db.serialTest).values({ value: raw<string>`get_value()` }),
    ).type.toRaiseError();
  });

  test('should insert using subquery', () => {
    expect(
      db
        .insertInto(db.foo)
        .values({ name: db.select(db.foo.name.concat(` 2`)).from(db.foo).limit(1) }),
    ).type.toBe<
      InsertQuery<
        Table<
          'foo',
          {
            id: Column<'id', 'foo', string, true, true, undefined>;
            createDate: Column<'createDate', 'foo', Date, true, true, undefined>;
            name: Column<'name', 'foo', string, true, false, undefined>;
            value: Column<'value', 'foo', number, false, false, undefined>;
          }
        >,
        number,
        {
          id: Column<'id', 'foo', string, true, true, undefined>;
          createDate: Column<'createDate', 'foo', Date, true, true, undefined>;
          name: Column<'name', 'foo', string, true, false, undefined>;
          value: Column<'value', 'foo', number, false, false, undefined>;
        }
      >
    >();
  });
});
