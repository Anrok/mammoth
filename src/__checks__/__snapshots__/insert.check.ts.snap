// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insert check should insert and await affect count (type) should match snapshot 1`] = `"number"`;

exports[`insert check should insert and returning count (type) should match snapshot 1`] = `"number"`;

exports[`insert check should insert by setting value to null (type) should match snapshot 1`] = `"InsertQuery<DefaultDbConfig, Table<DefaultDbConfig, \\"serialTest\\", { id: Column<DefaultDbConfig, \\"id\\", \\"serialTest\\", \\"Int4\\", true, true, undefined>; value: Column<DefaultDbConfig, \\"value\\", \\"serialTest\\", \\"Int4\\", false, false, undefined>; }>, number, { id: Column<DefaultDbConfig, \\"id\\", \\"serialTest\\", \\"Int4\\", true, true, undefined>; value: Column<DefaultDbConfig, \\"value\\", \\"serialTest\\", \\"Int4\\", false, false, undefined>; }>"`;

exports[`insert check should insert default column (type) should match snapshot 1`] = `"number"`;

exports[`insert check should insert multiple rows and returning count (type) should match snapshot 1`] = `"number"`;

exports[`insert check should insert using subquery (type) should match snapshot 1`] = `"InsertQuery<DefaultDbConfig, Table<DefaultDbConfig, \\"foo\\", { id: Column<DefaultDbConfig, \\"id\\", \\"foo\\", \\"Text\\", true, true, undefined>; createDate: Column<DefaultDbConfig, \\"createDate\\", \\"foo\\", \\"Date\\", true, true, undefined>; name: Column<DefaultDbConfig, \\"name\\", \\"foo\\", \\"Text\\", true, false, undefined>; value: Column<DefaultDbConfig, \\"value\\", \\"foo\\", \\"Int4\\", false, false, undefined>; }>, number, { id: Column<DefaultDbConfig, \\"id\\", \\"foo\\", \\"Text\\", true, true, undefined>; createDate: Column<DefaultDbConfig, \\"createDate\\", \\"foo\\", \\"Date\\", true, true, undefined>; name: Column<DefaultDbConfig, \\"name\\", \\"foo\\", \\"Text\\", true, false, undefined>; value: Column<DefaultDbConfig, \\"value\\", \\"foo\\", \\"Int4\\", false, false, undefined>; }>"`;

exports[`insert check should insert with expression of the not-null correct type (type) should match snapshot 1`] = `"InsertQuery<DefaultDbConfig, Table<DefaultDbConfig, \\"serialTest\\", { id: Column<DefaultDbConfig, \\"id\\", \\"serialTest\\", \\"Int4\\", true, true, undefined>; value: Column<DefaultDbConfig, \\"value\\", \\"serialTest\\", \\"Int4\\", false, false, undefined>; }>, number, { id: Column<DefaultDbConfig, \\"id\\", \\"serialTest\\", \\"Int4\\", true, true, undefined>; value: Column<DefaultDbConfig, \\"value\\", \\"serialTest\\", \\"Int4\\", false, false, undefined>; }>"`;

exports[`insert check should insert with expression of the nullable correct type (type) should match snapshot 1`] = `"InsertQuery<DefaultDbConfig, Table<DefaultDbConfig, \\"serialTest\\", { id: Column<DefaultDbConfig, \\"id\\", \\"serialTest\\", \\"Int4\\", true, true, undefined>; value: Column<DefaultDbConfig, \\"value\\", \\"serialTest\\", \\"Int4\\", false, false, undefined>; }>, number, { id: Column<DefaultDbConfig, \\"id\\", \\"serialTest\\", \\"Int4\\", true, true, undefined>; value: Column<DefaultDbConfig, \\"value\\", \\"serialTest\\", \\"Int4\\", false, false, undefined>; }>"`;

exports[`insert check should insert without explicit value for column serial (type) should match snapshot 1`] = `"InsertQuery<DefaultDbConfig, Table<DefaultDbConfig, \\"serialTest\\", { id: Column<DefaultDbConfig, \\"id\\", \\"serialTest\\", \\"Int4\\", true, true, undefined>; value: Column<DefaultDbConfig, \\"value\\", \\"serialTest\\", \\"Int4\\", false, false, undefined>; }>, number, { id: Column<DefaultDbConfig, \\"id\\", \\"serialTest\\", \\"Int4\\", true, true, undefined>; value: Column<DefaultDbConfig, \\"value\\", \\"serialTest\\", \\"Int4\\", false, false, undefined>; }>"`;

exports[`insert check should insert-returning and await rows (type) should match snapshot 1`] = `"{ name: string; }[]"`;

exports[`insert check should not insert by setting value to undefined (type) should match snapshot 1`] = `"undefined"`;

exports[`insert check should not insert invalid type in known column (type) should match snapshot 1`] = `"Type 'number' is not assignable to type 'string | TextExpression<DefaultDbConfig, \\"Text\\", boolean, string> | TextExpression<DefaultDbConfig, string, boolean, string> | Query<...>'."`;

exports[`insert check should not insert multiple rows with invalid colums (type) should match snapshot 1`] = `
"Type '{ name: string; asd: number; }' is not assignable to type '{ name: string | TextExpression<DefaultDbConfig, \\"Text\\", boolean, string> | TextExpression<DefaultDbConfig, string, boolean, string> | Query<...>; } & { ...; }'.
  Object literal may only specify known properties, and 'asd' does not exist in type '{ name: string | TextExpression<DefaultDbConfig, \\"Text\\", boolean, string> | TextExpression<DefaultDbConfig, string, boolean, string> | Query<...>; } & { ...; }'."
`;

exports[`insert check should not insert unknown column (type) should match snapshot 1`] = `
"Argument of type '{ name: string; asd: string; }' is not assignable to parameter of type '({ name: string | TextExpression<DefaultDbConfig, \\"Text\\", boolean, string> | TextExpression<DefaultDbConfig, string, boolean, string> | Query<...>; } & { ...; }) | ({ ...; } & { ...; })[]'.
  Object literal may only specify known properties, and 'asd' does not exist in type '({ name: string | TextExpression<DefaultDbConfig, \\"Text\\", boolean, string> | TextExpression<DefaultDbConfig, string, boolean, string> | Query<...>; } & { ...; }) | ({ ...; } & { ...; })[]'."
`;

exports[`insert check should not insert with wrong type of expression (type) should match snapshot 1`] = `
"Type 'TextExpression<DefaultDbConfig, string, false, \\"?column?\\">' is not assignable to type 'number | NumberExpression<DefaultDbConfig, \\"Int4\\", boolean, string> | NumberExpression<DefaultDbConfig, number, boolean, string> | Query<...>'.
  Type 'TextExpression<DefaultDbConfig, string, false, \\"?column?\\">' is missing the following properties from type 'NumberExpression<DefaultDbConfig, number, boolean, string>': plus, minus, multiply, divide, and 3 more."
`;
